switchboard/s3_util.py

```diff
@@ -0,0 +1,205 @@
+        self.user = user
+
+    def get_delivery(self):
+        return self.s3_delivery
+
+    def get_from_user(self):
+        return self.s3_delivery.from_user.user
+
+    def get_to_user(self):
+        return self.s3_delivery.to_user.user
+
+    def get_context(self):
+        from_user = self.get_from_user()
+        to_user = self.get_to_user()
+        bucket_name = self.s3_delivery.bucket.name
+        return {
+            's3_delivery_id': str(self.s3_delivery.id),
+            'from_name': '{} {}'.format(from_user.first_name, from_user.last_name),
+            'from_email': from_user.email,
+            'to_name': '{} {}'.format(to_user.first_name, to_user.last_name),
+            'to_email': to_user.email,
+            'project_title': bucket_name,
+            'project_url': 's://{}'.format(bucket_name)
+        }
+
+    def get_email_context(self, accept_url, process_type, reason, warning_message=''):
+        base_context = self.get_context()
+        return {
+            'project_name': base_context['project_title'],
+            'recipient_name': base_context['to_name'],
+            'recipient_email': base_context['to_email'],
+            'sender_email': base_context['from_email'],
+            'sender_name': base_context['from_name'],
+            'project_url': base_context['project_url'],
+            'accept_url': accept_url,
+            'type': process_type,  # accept or decline
+            'message': reason,  # decline reason
+            'user_message': self.s3_delivery.user_message,
+            'warning_message': warning_message,
+        }
+
+    @staticmethod
+    def from_s3_delivery_id(s3_delivery_id, user):
+        """
+        :param s3_delivery_id: int: S3Delivery id
+        :param user: django user
+        :return: S3DeliveryDetails
+        """
+        delivery = S3Delivery.objects.get(pk=s3_delivery_id)
+        return S3DeliveryDetails(delivery, user)
+
+    def get_action_template_text(self, action_name):
+        email_template = EmailTemplate.for_user(self.user, action_name)
+        if email_template:
+            return email_template.subject, email_template.body
+        else:
+            raise RuntimeError('No email template found')
+
+
+class S3Resource(object):
+    def __init__(self, s3_user):
+        session = boto3.session.Session(aws_access_key_id=s3_user.s3_id,
+                                        aws_secret_access_key=s3_user.credential.aws_secret_access_key)
+        self.s3 = session.resource('s3', endpoint_url=s3_user.endpoint.url)
+
+    def create_bucket(self, bucket_name):
+        bucket = self.s3.Bucket(bucket_name)
+        bucket.create()
+
+    def copy_bucket(self, source_bucket_name, destination_bucket_name):
+        # TODO: compare this on a large project vs aws cli sync
+        # http://boto3.readthedocs.io/en/latest/reference/services/s3.html#S3.Bucket.copy
+        # for many small files it sounds like this may be slower
```

Comment:
Based on the docs the aws cli python module sounds like it might be faster to perform this copy, but haven't tested this.

Reply:
For the first pass I am inclined to leave it as is.